pyomo/contrib/doe/measurements.py

```diff
@@ -25,252 +25,270 @@
+
+        if values:
+            # this dictionary keys are special set, values are its value
+            self.special_set_value = self._generate_dict(values)
+
+        if upper_bound:
+            if type(upper_bound) in [int, float]:
+                upper_bound = [upper_bound]*len(self.special_set)
+            self.upper_bound = self._generate_dict(upper_bound)
+            
+        if lower_bound:
+            if type(lower_bound) in [int, float]:
+                lower_bound = [lower_bound]*len(self.special_set)
+            self.lower_bound = self._generate_dict(lower_bound)
+
+        return self.special_set
+    
+    def update_values(self, values):
         """
-        Generate a dictionary, keys are the variable names, values are the indexes of this variable.
-        For e.g., name_and_index = {'C': ['CA', 'CB', 'CC']}
+        Used for updating values
+
         Parameters
         ---------
-        all_info: a dictionary, keys are measurement variable names,
-                values are a dictionary, keys are its extra index, values are its measuring time points
-                values are a list of measuring time point if there is no extra index for this measurement
-            Note: all_info can be the self.measurement_all_info, but does not have to be it.
+         values: a ``list`` containing values which has the same shape of flattened variables 
+            default choice is None, means there is no give nvalues 
         """
-        measurement_name = list(all_info.keys())
-        # a list of measurement extra indexes
-        measurement_extra_index = []
-        # check if the measurement has extra indexes
-        for i in measurement_name:
-            if type(all_info[i]) is dict:
-                index_list = list(all_info[i].keys())
-                measurement_extra_index.append(index_list)
-            elif type(all_info[i]) is list:
-                measurement_extra_index.append(None)
-        # a dictionary, keys are measurement names, values are a list of extra indexes
-        self.name_and_index = dict(zip(measurement_name, measurement_extra_index))
-
-    def _generate_flatten_name(self, measure_name_and_index):
+        self.special_set_value = self._generate_dict(values)
+        
+
+    def _generate_dict(self, values):
         """
-        Generate measurement flattened names
-        Parameters
-        ----------
-        measure_name_and_index: a dictionary, keys are measurement names, values are lists of extra indexes
+        Given a list of values, return a dictionary, keys are special set names. 
+        """
+        value_map = {}
+        for i in range(len(values)):
+            value_map[self.special_set[i]] = values[i]
 
-        Returns
-        ------
-        jac_involved_name: a list of flattened measurement names
+        return value_map
+
+
+    def _add_elements(self, var_name, extra_index=None, time_index=None):
```

Comment:
Change this function name to `_add_variables`.

The biggest change is going from `var_name` being a list with all variables and calling this function once to making var_name a string and calling this function multiple times, once for each different variable. This change will make the API much clearer. As part of the change, we need to think through all of the implications of calling this function multiple times. In several places, we implicitly assume this function is called only once.